californium-core/src/main/java/org/eclipse/californium/core/network/stack/TCPBlockwiseLayer.java

```diff
@@ -0,0 +1,394 @@
+						block1.getNum());
+
+				// Remember block to acknowledge. TODO: We might make
+				// this a
+				// boolean flag in status.
+				exchange.setBlock1ToAck(block1);
+
+				// Assemble and deliver
+				Request assembled = new Request(request.getCode());
+				assembled.setSenderIdentity(request.getSenderIdentity());
+				status.assembleMessage(assembled);
+
+				assembled.setMID(request.getMID());
+				assembled.setToken(request.getToken());
+
+				assembled.getOptions().setBlock2(request.getOptions().getBlock2());
+
+				clearBlock1Status(key);
+
+				exchange.setRequest(assembled);
+				upper().receiveRequest(exchange, assembled);
+			}
+		} else {
+			super.handleInboundBlockwiseUpload(exchange, request);
+		}
+
+	}
+
+	/**
+	 * Handles the request for next block from the peer. Handling is done here
+	 * only for SZX = 7 all other handling is done in the BlockwikseLayer.
+	 */
+	protected void handleInboundRequestForNextBlock(final Exchange exchange, final Request request, final KeyUri key,
+			final Block2BlockwiseStatus status) {
+
+		BlockOption block2 = request.getOptions().getBlock2();
+		if ((block2 != null) && (block2.getSzx() == BERT_SZX)) {
+
+			synchronized (status) {
+
+				ByteBuffer responseBlockBuilder = ByteBuffer.allocate(bertStepSize * BERT_INT_BLOCK_SIZE);
+				int blockNum = block2.getNum();
+				int currentNum = blockNum;
+				boolean hasNextBlock = true;
+				Response block = null;
+
+				for (int i = 0; ((i < bertStepSize) && (hasNextBlock)); i++) {
+					block = status.getNextResponseBlock(block2);
+					hasNextBlock = block.getOptions().getBlock2().isM();
+					currentNum = currentNum + 1;
+					block2 = new BlockOption(BERT_SZX, hasNextBlock, currentNum);
+					responseBlockBuilder.put(block.getPayload());
+				}
+
+				if (status.isComplete()) {
+					// clean up blockwise status
+					LOGGER.log(Level.FINE, "peer has requested last block of blockwise transfer: {0}", status);
+					clearBlock2Status(key);
+				} else {
+					LOGGER.log(Level.FINE, "peer has requested intermediary block of blockwise transfer: {0}", status);
+				}
+
+				block.setPayload(getBody(responseBlockBuilder));
+				boolean m = block.getOptions().getBlock2().isM();
+				block.getOptions().setBlock2(BERT_SZX, m, blockNum);
+				exchange.setCurrentResponse(block);
+				lower().sendResponse(exchange, block);
+			}
+		} else {
+			super.handleInboundRequestForNextBlock(exchange, request, key, status);
+		}
+	}
+
+	/**
+	 * Send request to peers.Handling is done here only when BERT option is
+	 * enabled otherwise handling is done in the BlockwikseLayer.
+	 */
+	public void sendRequest(final Exchange exchange, final Request request) {
+
+		Request firstRequest = request;
+		Request requestToSend;
+		if (isBertEnabled) {
+
+			if (isTransparentBlockwiseHandlingEnabled()) {
+
+				BlockOption block2 = request.getOptions().getBlock2();
+				if (block2 != null && block2.getNum() > 0) {
+					isBertEnabled = false;
```

Comment:
What should be the behaviour in case  BERT is enabled ?  Should BERT be disabled only for the Current Message ? or should BERT not be disabled at all ?